File: hackerrank/factorial.js
function factorial(n){
    // 1. instantiate scope variables
    var result = 1;

    // 3a. helper function
    function multiplyInt(count){
        
        // 4. base case
        if(count > n){ return; }
        
        // 5. recursive case
        result *= count;
        multiplyInt(count + 1);
    }

    // 3b. then invoke helper function 
    multiplyInt(1);

    // 2. return result
    return result;
}

File: hackerrank/factorial.py
def factorial(n):
    # instantiate scope variables as collections (e.g., list)
    result = [1]
    # helper method
    def multiply_int(count):
        # base case
        if(count > n):
            return
        #recursive case
        result[0] = result[0] * count # <-- accessing an element in outer scope
        multiply_int(count + 1)
    # invoke helper method with initial input parameters
    multiply_int(1)
    #return result
    return result[0]